File: LeetCode/C#/0206-ReverseLinkedList.cs
/*

Approach:
1. We will use three pointers, prev, curr and temp.
2. We will iterate through the linked list and reverse the pointers.
3. We will return the prev pointer as the head of the reversed linked list.

Time Complexity: O(n)
Space Complexity: O(1)

*/
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode ReverseList(ListNode head) {

        ListNode prev, curr, temp;  // prev is the previous node, curr is the current node and temp is the temporary node.
        curr = head;                // curr is the head of the linked list.
        prev = null;                // prev is null initially.
        while(curr != null) {       // iterate through the linked list.
            temp = curr.next;       // store the next node in temp.
            curr.next = prev;       // reverse the pointer.
            prev = curr;            // move the prev pointer to the current node.
            curr = temp;            //  move the current pointer to the next node.
        }
        return prev;                //  return the prev pointer as the head of the reversed linked list.
    }
}

File: LeetCode/Go/0206-ReverseLinkedList.go
/*
Solution: 
Difficulty: Easy
Approach:
Tags: Array, Linked List, String
1. We will use three pointers, prev, curr and temp.
2. We will iterate through the linked list and reverse the pointers.
3. We will return the prev pointer as the head of the reversed linked list.
Space Complexity: O(1)

Time Complexity: O(n)
Space Complexity: O(1)
*/

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func reverseList(head *ListNode) *ListNode {
    var curr *ListNode;		// Declare the current node
    var prev *ListNode;		// Declare the previous node
    var temp *ListNode;		// Declare the temporary node

    curr= head				// Initialize the current node
    prev= nil				// Initialize the previous node
    for curr != nil {		// Iterate through the linked list
        temp= curr.Next		// Store the next node
        curr.Next=prev		// Reverse the current node
        prev = curr			// Move the previous node
        curr= temp			// Move the current node
    }
    return prev;			// Return the reversed linked list
}